### PR TITLE
Nxinx Kubernetes-Ingress Virtualserver support

### PR DESCRIPTION
**Pull request by Kaezon**
This PR will enable support for [nginxinc kubernetes-ingress](https://github.com/nginxinc/kubernetes-ingress) VirtualServer CRs.
In the process of working on this, I updated a couple test resources by using `helm template` to render newer manifests into `ingress.yaml` and `metallb.yaml`

I think #65 also made updates to `metallb.yaml`, so this might not be necessary anymore.
`ingress.yaml` contained a manifest for `kubernetes/ingress-nginx`. I replaced it with a render of `nginxinc/kubernetes-ingress`. They are both based on the open source edition of nginx, and have a large functionality overlap; however, this deployment includes the extra functionality I wanted to integrate. I figured it would be ok to replace the original since it was also nginx, but I could move the newer version into the integration test directory and roll back if that would be preferable.

More info on the differences between the ingress controllers: [here](https://www.nginx.com/blog/guide-to-choosing-ingress-controller-part-4-nginx-ingress-controller-options/)

Changelog:
- Added VirtualServer information to README.md
- Added test resources to Tiltfile
- Added `github.com/nginxinc/kubernetes-ingress v1.12.3` to `go.mod`
- Updated `go.sum`
- Added VirtualServer to `orderedResources` in `gateway.go` with priority over Ingress objects
- Implemented VirtualServer listing, watching, indexing, and lookup functionality in `kubernetes.go`
- Added VirtualServer CRD check to `kubernetes.go`
- Added nginx ClientSet to `KubeController` struct
- Updated unit tests
  - gateway_test.go
  - kubernetes_test.go
- Updated test resources
  - Rendered nginx kubernetes-ingress helm chart into `test/ingress.yaml`
  - Rendered metallb helm chart into `test/metallb.yaml`
  - Added `test/nginxinc-kubernetes-ingress/resources.yaml`

### CODE REVIEW

File: kubernetes.go [hunk 1]
```diff
@@ -169,6 +195,18 @@ func existGatewayCRDs(ctx context.Context, c *gatewayClient.Clientset) bool {
 	panic(err)
 }
 
+func existVirtualServerCRDs(ctx context.Context, c *k8s_nginx.Clientset) bool {
+	_, err := c.K8sV1().VirtualServers("").List(ctx, metav1.ListOptions{})
+	if meta.IsNoMatchError(err) || runtime.IsNotRegisteredError(err) || errors.IsNotFound(err) {
+		log.Infof("VirtualServer CRDs are not found. Not syncing VirtualServer resources.")
+		return false
+	}
```
Comment (networkop) [inline kubernetes.go:203]: can you update this function with https://github.com/ori-edge/k8s_gateway/pull/71?
can you also see if it's possible to factor out this entire error check into a single function?

**Review by networkop — CHANGES_REQUESTED** [review]
overall looks good. 
I still would prefer to keep the original nginx-ingress controller in place, just because of how common it is. ( I realise that processing of Ingress resources is exactly the same).
Would it be possible to have both of them co-exist, just separate them into different classes and make sure they don't try to manage each other's resources?

File: gateway.go [hunk 2]
```diff
@@ -23,6 +23,9 @@ var orderedResources = []*resourceWithIndex{
 	{
 		name: "HTTPRoute",
 	},
+	{
+		name: "VirtualServer",
```
Comment (networkop) [inline gateway.go:27]: can you also update this with this little change https://github.com/ori-edge/k8s_gateway/pull/72

File: kubernetes.go (hunk 1 above)
Comment (Kaezon) [reply kubernetes.go:208]: Updated.
I quickly looked at combining the functions, but the key problem is that each uses a different implementation of `ClientSet`. We could pass both `ClientSet` objects into a single function and return an ordered set of bools to cover each CRD.
Would that be acceptable?

File: gateway.go (hunk 2 above)
Comment (Kaezon) [reply gateway.go]: Done

File: kubernetes.go (hunk 1 above)
Comment (networkop) [reply kubernetes.go:208]: I was thinking more along the lines of the error processing function...
so from the existXCRDs() we would call `checkErr(err)` 
and `func checkErr(err error)` would do the if checks and print statements. 
maybe it doesn't make sense, not sure, just an idea, not a req

Comment (Kaezon) [reply kubernetes.go:208]: No, that makes sense. That should be simple enough.

Comment (Kaezon) [reply kubernetes.go:205]: added a common error process

**Comment by Kaezon** [comment]
> overall looks good. I still would prefer to keep the original nginx-ingress controller in place, just because of how common it is. ( I realise that processing of Ingress resources is exactly the same). Would it be possible to have both of them co-exist, just separate them into different classes and make sure they don't try to manage each other's resources?

I moved the nginxinc ingress into the namespaced directory and configured it with the `ingressClassName` of `nginxinc`, so we shouldn't have any collisions.

**Comment by networkop** [comment]
I think this looks great. Thanks for contribution @Kaezon

**Comment by networkop** [comment]
@Kaezon I assume that https://github.com/ori-edge/k8s_gateway/blob/master/test/nginxinc-kubernetes-ingress/values.yaml is what you'd used to `helm generate` the `ingress.yaml` , right? 
I'm planning to phase out these static manifests and replace them with helm actions in tilt, so wanted to make sure I've got all I need to install the nginx controller.

**Comment by Kaezon** [comment]
@networkop Correct. I saw the PR for moving to Helm and figured I may as well make an actual values file and leave it there.

**Pull request merged**